File: ios/browser/brave_wallet/swap_controller_factory.cc
/* Copyright (c) 2021 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "brave/ios/browser/brave_wallet/swap_controller_factory.h"

#include "brave/components/brave_wallet/browser/swap_controller.h"
#include "components/keyed_service/core/keyed_service.h"
#include "components/keyed_service/ios/browser_state_dependency_manager.h"
#include "ios/chrome/browser/browser_state/browser_state_otr_helper.h"
#include "ios/chrome/browser/browser_state/chrome_browser_state.h"
#include "services/network/public/cpp/shared_url_loader_factory.h"

namespace brave_wallet {

// static
mojom::SwapController* SwapControllerFactory::GetForBrowserState(
    ChromeBrowserState* browser_state) {
  return static_cast<SwapController*>(
      GetInstance()->GetServiceForBrowserState(browser_state, true));
}

// static
SwapControllerFactory* SwapControllerFactory::GetInstance() {
  return base::Singleton<SwapControllerFactory>::get();
}

SwapControllerFactory::SwapControllerFactory()
    : BrowserStateKeyedServiceFactory(
          "SwapController",
          BrowserStateDependencyManager::GetInstance()) {}

SwapControllerFactory::~SwapControllerFactory() = default;

std::unique_ptr<KeyedService> SwapControllerFactory::BuildServiceInstanceFor(
    web::BrowserState* context) const {
  auto* browser_state = ChromeBrowserState::FromBrowserState(context);
  std::unique_ptr<SwapController> swap_controller(
      new SwapController(browser_state->GetSharedURLLoaderFactory()));
  return swap_controller;
}

bool SwapControllerFactory::ServiceIsNULLWhileTesting() const {
  return true;
}

web::BrowserState* SwapControllerFactory::GetBrowserStateToUse(
    web::BrowserState* context) const {
  return GetBrowserStateRedirectedInIncognito(context);
}

}  // namespace brave_wallet

File: android/java/org/chromium/chrome/browser/crypto_wallet/SwapControllerFactory.java
/* Copyright (c) 2021 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser.crypto_wallet;

import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.brave_wallet.mojom.SwapController;
import org.chromium.mojo.bindings.ConnectionErrorHandler;
import org.chromium.mojo.bindings.Interface;
import org.chromium.mojo.bindings.Interface.Proxy.Handler;
import org.chromium.mojo.system.MessagePipeHandle;
import org.chromium.mojo.system.impl.CoreImpl;

@JNINamespace("chrome::android")
public class SwapControllerFactory {
    private static final Object sLock = new Object();
    private static SwapControllerFactory sInstance;

    public static SwapControllerFactory getInstance() {
        synchronized (sLock) {
            if (sInstance == null) {
                sInstance = new SwapControllerFactory();
            }
        }
        return sInstance;
    }

    private SwapControllerFactory() {}

    public SwapController getSwapController(ConnectionErrorHandler connectionErrorHandler) {
        int nativeHandle = SwapControllerFactoryJni.get().getInterfaceToSwapController();
        MessagePipeHandle handle = wrapNativeHandle(nativeHandle);
        SwapController swapController = SwapController.MANAGER.attachProxy(handle, 0);
        Handler handler = ((Interface.Proxy) swapController).getProxyHandler();
        handler.setErrorHandler(connectionErrorHandler);

        return swapController;
    }

    private MessagePipeHandle wrapNativeHandle(int nativeHandle) {
        return CoreImpl.getInstance().acquireNativeHandle(nativeHandle).toMessagePipeHandle();
    }

    @NativeMethods
    interface Natives {
        int getInterfaceToSwapController();
    }
}
